### PR TITLE
[MM-47843] Work Template Modal (read-only)

### PR DESCRIPTION
**Pull request by JulienTant**
#### Summary
The goal of this PR is to provide a first version of a "dev-ready" work template modal that is feature flag gated (see server PR).

***When reviewing*** keep in mind that we're looking to have this merged in a "dev-ready" state and not "prod-ready" ~ So the "dev" parts (inject fake data in the store etc...) are "ok" and expected to stay but please bring up any quality concern!

*How to access it*

1) You must run the server with the env `export MM_FEATUREFLAGS_WorkTemplate='true'`. 
2) There's a temporary entry called "Work Template" in our "Application" menu (Top Left of your screen). Again, only visible if the Feature Flag is on.

*Want more details? Sure!*

The idea is that each page of the modal content is 100% generated by the work templates that come from the redux store. In this first iteration, the work templates are hardcoded in a file and injected into the redux store to "simulate" what we would get from the server in the near future. I tried to prefix (or include) the keyword "dev" and or "todo" in every method/structure/file that is meant to be removed eventually.

You can go as far as your want in the process, the "final" result when using "Quick Create" or "Create" from the last modal page will trigger an alert to let us know what's going to happen. Long term idea is that we'll trigger a request to the server and the server will take create of creating everything that's necessary.

Here's a non-complete list of what is missing in this version on the UI side (and that won't be added in this PR):

- The types are not yet "complete".
- Create a definitive entry in the "application" menu
- Create an entry in the "onboarding" window
- The integrations (plugin) part has been completely left out.
- Getting templates from the backend rather than from the a dev file in the frontend

**This branch accidentally turned into a feature branch, to avoid the challenges associated with a feature branch, it'll be merged in master and we'll iterate over master**. Everything is behind a feature flag, so this is a safe operation.

#### Ticket Link
[mattermost.atlassian.net/browse/MM-47843](https://mattermost.atlassian.net/browse/MM-47843)

#### Related Pull Requests
- Has server changes https://github.com/mattermost/mattermost-server/pull/21553

#### Screenshots
Go see for yourself :D 

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by JulienTant** [comment]
Alright I've setup the Spinwick with the FeatureFlag **on**!

**Review by hmhealey — COMMENTED** [review]
Not a full review since I was just stopping by to see how far along this was, but I wanted to mention a few things I noticed.

Also, I also really like the dev data stuff you're doing here. I've always liked that sort of demo, but we don't tend to do it very often even though that usually means that we have frontend stuff blocked on backend changes.

File: components/worktemplates/components/modal.tsx
Comment (julmondragon) [inline components/worktemplates/components/modal.tsx:9]: Should take into account tablet and mobile [screen width](https://github.com/mattermost/mattermost-webapp/blob/master/utils/constants.tsx#L1378-L1380).

File: components/worktemplates/components/preview.tsx
Comment (julmondragon) [inline components/worktemplates/components/preview.tsx:25]: import `useState`

File: components/worktemplates/components/preview.tsx
Comment (julmondragon) [inline components/worktemplates/components/preview.tsx:30]: What about a single `for` to set all the required data (channels, boards, etc) ?

File: components/worktemplates/components/preview.tsx
Comment (julmondragon) [inline components/worktemplates/components/preview.tsx:37]: intl ?

File: components/worktemplates/components/preview/accordion.tsx
Comment (julmondragon) [inline components/worktemplates/components/preview/accordion.tsx:1]: Could this customization just be available in the original component? Maybe controlled through props?

File: components/worktemplates/index.tsx
Comment (julmondragon) [inline components/worktemplates/index.tsx:56]: `state` could be confusing. Type ?

**Review by julmondragon — COMMENTED** [review]
First round of review done. Great work here @JulienTant

File: packages/mattermost-redux/src/reducers/entities/worktemplates.ts
Comment (JulienTant) [reply packages/mattermost-redux/src/reducers/entities/worktemplates.ts:4]: Oh wow, I wish we could make this more "obvious" somehow 🤔

File: packages/types/src/utilities.ts
Comment (JulienTant) [reply packages/types/src/utilities.ts:31]: I don't have a use case, I'll drop it.

File: components/worktemplates/components/menu/use_case.tsx
Comment (JulienTant) [reply components/worktemplates/components/menu/use_case.tsx:31]: oops yeah very french way to indicate a count lol

File: components/worktemplates/components/menu/use_case.tsx
Comment (JulienTant) [reply components/worktemplates/components/menu/use_case.tsx:9]: I didn't know the existence of this file, that being said there's no defined button-size that matches this one, and no "colorset" that matches either.

File: components/worktemplates/components/modal.tsx
Comment (JulienTant) [reply components/worktemplates/components/modal.tsx:9]: It works fine on desktop and tablet, but I'm unsure how to address the feature on mobile.  @annelieseklein is this something that's been considered?

File: components/worktemplates/components/preview/accordion.tsx
Comment (JulienTant) [reply components/worktemplates/components/preview/accordion.tsx:1]: I'm unsure if applying this style to the main component behind a prop is bringing any benefits 🤔 It's just forcing the generic component to be responsible for the specific style of this modal

File: components/worktemplates/components/customize.tsx
Comment (JulienTant) [reply components/worktemplates/components/customize.tsx:40]: So that is what I did first, but:

- My HTML ended up being bigger (had to set `useLegend` to false)
- I ended up writing more CSS because it comes with a fieldset that contains border and background colors and padding

So in the end I felt like I was fighting the `<Input>` field more than I was using it 😅 .

File: components/worktemplates/index.tsx
Comment (mhd-sln) [inline components/worktemplates/index.tsx:78]: strict equality not needed

File: packages/types/src/worktemplates.ts
Comment (pvev) [inline packages/types/src/worktemplates.ts:37]: these interfaces have equal body (Channel, Board, Playbook), you could declare one boilerplate and then use type (under the hood interfaces and types are the same)

File: components/worktemplates/index.tsx
Comment (pvev) [inline components/worktemplates/index.tsx:12]: nit: group imports by context

File: components/worktemplates/components/modal_body_with_illustration.tsx
Comment (pvev) [inline components/worktemplates/components/modal_body_with_illustration.tsx:19]: nit: this style here looks weird, you can set it in the styled component too

File: components/worktemplates/components/menu/use_case.tsx
Comment (pvev) [inline components/worktemplates/components/menu/use_case.tsx:132]: nit: unnecessary blank line

File: components/worktemplates/components/menu/use_case.tsx
Comment (pvev) [inline components/worktemplates/components/menu/use_case.tsx:115]: should this color be themeable? (not sure if that is a word)

File: components/worktemplates/components/menu/use_case.tsx
Comment (pvev) [inline components/worktemplates/components/menu/use_case.tsx:44]: nit: no need for checking the greater than 0, after 1 all values are truthy

File: components/worktemplates/components/menu.tsx
Comment (pvev) [inline components/worktemplates/components/menu.tsx:82]: useSelector can potentially return undefined, for defensive programming I would set this as `categories?.length`

File: components/worktemplates/components/menu.tsx
Comment (pvev) [inline components/worktemplates/components/menu.tsx:24]: nit: unnecessary blank line, 🙄 styled components 🙄

File: components/worktemplates/components/menu.tsx
Comment (pvev) [inline components/worktemplates/components/menu.tsx:13]: nit: try to organize imports by context, it is really helpful when making code archeology

File: components/worktemplates/components/customize.tsx
Comment (pvev) [inline components/worktemplates/components/customize.tsx:89]: nit: there is an error with the formatting here, but the linter is not able to catch this kind of problems when using styled components.  😢 .. my very very personal opinion/preference, keep using scss modules

File: components/worktemplates/components/customize.tsx
Comment (pvev) [inline components/worktemplates/components/customize.tsx:63]: nit: I know this is used this way every where, but a good practice is to avoid the use of magic strings/characters, for these cases a constant/enum is more semantic and the code is more organized

File: components/worktemplates/components/customize.tsx
Comment (pvev) [inline components/worktemplates/components/customize.tsx:50]: I guess 'Test' is temporal

File: components/worktemplates/components/customize.tsx
Comment (pvev) [inline components/worktemplates/components/customize.tsx:34]: ❤️ love the use of semantic html, but not sure if this is the place for strong or should be just <b>

File: components/worktemplates/components/customize.tsx
Comment (pvev) [inline components/worktemplates/components/customize.tsx:9]: nit: keep imports grouped by context, in this case the 'components' one together and the '@mattermost' in a separate lines

File: components/global_header/left_controls/product_menu/product_menu_list/product_menu_list.tsx
Comment (pvev) [inline components/global_header/left_controls/product_menu/product_menu_list/product_menu_list.tsx:146]: should this be translated?

File: components/common/accordion/accordion.tsx
Comment (pvev) [inline components/common/accordion/accordion.tsx:32]: question, wouldn't make sense to receive the indexes of the items we would like to be opened on load? in this case you would just send `[0]`

**Review by pvev — APPROVED** [review]
Great initial work @JulienTant , just a couple of nits and couple of questions, but so far looking really good. Thanks! 🚀

File: packages/types/src/worktemplates.ts
Comment (JulienTant) [reply packages/types/src/worktemplates.ts:37]: This is a temporary situation, we'll have more fields in each of them once the API is plugged in.

File: components/worktemplates/index.tsx
Comment (JulienTant) [reply components/worktemplates/index.tsx:12]: Not sure what do you mean by context?

File: components/worktemplates/components/menu/use_case.tsx
Comment (JulienTant) [reply components/worktemplates/components/menu/use_case.tsx:115]: I'll defer that question to @annelieseklein ^^ this is the very first time we use this color.

File: components/worktemplates/components/menu.tsx
Comment (JulienTant) [reply components/worktemplates/components/menu.tsx:24]: you meant 

> styled components ❤️

File: components/worktemplates/components/customize.tsx
Comment (JulienTant) [reply components/worktemplates/components/customize.tsx:81]: To your point, it might be worth checking https://styled-components.com/docs/tooling#stylelint to allow stylelint to parse styled components 🤔 I surely won't go through this with this PR, it is big enough as it is without fixing tons of random files when this is added haha! But this is something that could be explored.

File: components/worktemplates/components/customize.tsx
Comment (JulienTant) [reply components/worktemplates/components/customize.tsx:55]: Absolutely right, it fell off my radar!

File: components/worktemplates/components/customize.tsx
Comment (JulienTant) [reply components/worktemplates/components/customize.tsx]: oops forgot to push the last changed it got removed

File: components/global_header/left_controls/product_menu/product_menu_list/product_menu_list.tsx
Comment (JulienTant) [reply components/global_header/left_controls/product_menu/product_menu_list/product_menu_list.tsx:146]: yes!

File: components/common/accordion/accordion.tsx
Comment (JulienTant) [reply components/common/accordion/accordion.tsx:32]: That's actually what I did first and I reverted to this because it made more semantic sense? I had my accordion like:

```ts
<Accordion 
   //...
   openOnFirstDraw={[0]}
   //...
/>
```

But realized that most of the time we'll want the first element to be open or everything close by default, and a bool was more elegant than a `[0]` as an argument. I guess I'm 3/5 on that but if there's other use cases in the near future that needs such a change it would make sense to allow it

Comment (pvev) [reply components/common/accordion/accordion.tsx:32]: ok, the point is valid, but if you are making the change to the component I think receiving the array instead of a boolean makes more sense, you can always use a semantic constant name like `OPEN_FIRST_ELEMENT = 0; ... [OPEN_FIRST_ELEMENT] ` and will be as explanative as the boolean value. My point is, at the end you are making the same change, one gets limited to the first element the other one is more extendable to open other elements.

File: components/worktemplates/components/customize.tsx
Comment (pvev) [reply components/worktemplates/components/customize.tsx:89]: or we could stop using styled.components 🙈

File: components/worktemplates/index.tsx
Comment (pvev) [reply components/worktemplates/index.tsx:10]: all the ones form 'components' in one group, the ones from '@mattermost' another group, separated by blank lines, again, this is a nit ;)

File: packages/types/src/worktemplates.ts
Comment (pvev) [reply packages/types/src/worktemplates.ts:37]: R

File: components/worktemplates/components/menu.tsx
Comment (hmhealey) [inline components/worktemplates/components/menu.tsx:44]: There's some spots here that I think need to match the user's current theme, not the denim theme since I think these appear in the regular app and not in an unthemed page (like the System Console)

File: components/worktemplates/components/menu/use_case.tsx
Comment (hmhealey) [inline components/worktemplates/components/menu/use_case.tsx:51]: I know I commented on using plurals earlier in this PR, but I didn't realize that you could add `#` into them to add the actual count. That's so much easier than adding it separately

File: components/worktemplates/index.tsx
Comment (hmhealey) [reply components/worktemplates/index.tsx:10]: We have an ESLint configuration that does some amount of this, but we really need to update it at some point to do the rest of that for us

File: packages/mattermost-redux/src/action_types/worktemplates.ts
Comment (hmhealey) [inline packages/mattermost-redux/src/action_types/worktemplates.ts:7]: Nit: This should probably be two words like `RECEIVED_DEV_WORK_TEMPLATES`

File: packages/mattermost-redux/src/action_types/worktemplates.ts
Comment (hmhealey) [inline packages/mattermost-redux/src/action_types/worktemplates.ts:1]: In a similar vein, these file and folder names should spell work templates as `work_templates`

File: packages/types/src/utilities.ts
Comment (hmhealey) [inline packages/types/src/utilities.ts:25]: What's this supposed to be used for? I'm not sure what `RequireOnlyOne` would mean in this context

File: utils/constants.tsx
Comment (hmhealey) [inline utils/constants.tsx:434]: Work template should also be two separate words here as well

File: components/worktemplates/components/menu/use_case.tsx
Comment (JulienTant) [reply components/worktemplates/components/menu/use_case.tsx:51]: I could take credit but GitHub Autopilot suggested this for me :P

File: components/worktemplates/components/menu.tsx
Comment (JulienTant) [reply components/worktemplates/components/menu.tsx:44]: the variable comes from `_css_variables.scss` and its value has no other equivalent, it also appears to be themed (bg color of the modal on page 2 and 3):

image

File: packages/types/src/utilities.ts
Comment (JulienTant) [reply packages/types/src/utilities.ts:25]:
```ts
type Test RequireOnlyOne<{
   Obj1? any;
   Obj2? any;
   Obj3? any;
   Obj4? any;
}>
```

Only one of the Obj* can be set on a variable of test Test. Note that this already exist in mattermost-webapp @ `/utils/conditional_types.ts` but was not accessible from the packages.

File: utils/constants.tsx
Comment (JulienTant) [reply utils/constants.tsx:435]: if this is really important, I'll make a ticket to do that later (but before prod), as we have tickets ongoing based on this branch and changing request/changing those values, I would rather avoid conflict right now and go through a renaming session later (I would also have to rename a bunch on things in the backend)

File: packages/mattermost-redux/src/action_types/worktemplates.ts
Comment (marianunez) [reply packages/mattermost-redux/src/action_types/worktemplates.ts:1]: +1 on this

File: packages/types/src/worktemplates.ts
Comment (marianunez) [reply packages/types/src/worktemplates.ts:37]: Let's leave a TODO comment on that so that there is more context for others that stumble into this later

File: utils/constants.tsx
Comment (marianunez) [reply utils/constants.tsx:435]: I would +1 on the renaming - given that the other PRs have merged here, should we rename it before merge?

Comment (hmhealey) [reply utils/constants.tsx:435]: I think we should rename it first or else we're going to end up with a mix of code using one and code using the other

File: packages/types/src/utilities.ts
Comment (JulienTant) [reply packages/types/src/utilities.ts:25]: Yes indeed:

```ts
// clone of packages/mattermost-redux/src/types/actions.ts used for this test
export type ActionResult<Data = any, Error = any> = {
    data?: Data;
    error?: Error;
};

type ActionResultBis = RequireOnlyOne<ActionResult>;

const testOriginal: ActionResult = {
    data: 'test',
    error: 'test',
}; // <-- no error

const testWithRequireOnlyOne: ActionResultBis = {
    data: 'test',
    error: 'test',
}; // <-- error
```

Comment (JulienTant) [reply packages/types/src/utilities.ts:25]: > In that case, would you mind just getting rid of utils/conditional_types and moving the other type here as well?

done!

File: utils/constants.tsx
Comment (JulienTant) [reply utils/constants.tsx:435]: Well that was a fun refactoring :P 35 files changed, we're all good now!

**Comment by JulienTant** [comment]
This branch accidentally turned into a feature branch, to avoid the challenges associated with a feature branch, it'll be merged in master and we'll iterate over master. Everything is behind a feature flag, so this is a safe operation.

File: utils/constants.tsx
Comment (marianunez) [reply utils/constants.tsx]: Fun stuff for sure 💯 🙃

**Pull request merged**
